Journalist Can Create Articles And Place Them In Appropriate Category

angiecstn opened the pull request:
PT Link: https://www.pivotaltracker.com/story/show/161219985

* Adding attribute byline to Article model
* Adding create and new functionality in article controller
* Adding load categories action in application controller 
* Add requirements for category in create method
* Adding admin form in view (new.html.haml)
* Configuring resources in routes
* Refactoring to have html in haml

![screen shot 2018-10-19 at 16 56 54](https://user-images.githubusercontent.com/33787018/47226111-053f0d80-d3c0-11e8-804e-7848de3f26d7.png)

angiecstn committed:
adding create method for articles

RhonGabriel committed:
created new view, added form, test green

angiecstn committed:
adding features and steps for article form, adding category checkbox and submit button in html

angiecstn committed:
changing new.html.erb to haml

angiecstn committed:
adding route for create article method, adding flash message for article created both successfully and unsuccessfully

angiecstn committed:
[troubleshooting] adding byline attribute to article model, changing checkbox to scrolldown, updating test for it

angiecstn committed:
fixing chrome gem troubleshooting

angiecstn committed:
refactoring basic steps and feature for submit and scrolldown, fixing flash notice in controller, all test passing

angiecstn committed:
refactoring

ESteinman commented on features/step_definitions/basic_steps.rb:
```diff
@@ -1,8 +1,3 @@
-# frozen_string_literal: true
-Then "stop" do
```
You only need to remove the "Then stop" in the feature, not the actual step-definition, you can keep this ;)

angiecstn committed:
adding application.html.haml instead of application.html.erb

angiecstn committed:
Merge branch 'development' into create-articles

RhonGabriel committed:
fixed indentation application html

rupertlion commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,8 @@
+    
+    private
+    
+    def load_categories
```
Just for me to understand: why is this load_categories in here and why is it private?

angiecstn replied:
Because we have then added the load_categories to the article controller as before action, so that every time a new article is created in the application before it happens we are actually loading the categories.

rupertlion commented on app/controllers/articles_controller.rb:
```diff
@@ -1,9 +1,33 @@
+      end
+  end
+
+
```
Watch your white space

rupertlion commented on app/views/layouts/application.html.haml:
```diff
@@ -1,12 +1,18 @@
     = javascript_include_tag 'application', 'data-turbolinks-track': 'reload'
   %body
-    = yield
+  create-articles
```
what is this and why is it here and also on line 4?

angiecstn replied:
It is needed because when we save or try to save a new article we are rendering a flash alert in the layout. I fixed the repetition issue.

rupertlion commented on config/routes.rb:
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
-  root controller: :articles, action: :index
+  root controller: :articles, action: :index 
+  
```
whitespace

rupertlion commented on app/controllers/articles_controller.rb:
```diff
@@ -1,9 +1,33 @@
+
+
+  def new
+    @article = Article.new
```
Could this not be inside the create function?

angiecstn replied:
The new method is needed for the if / else statement if the article cannot be saved and we call on  it inside of the create method, if placed inside of the create method it doesn't find the load_categories so also the collection_select in the new.html.haml

rupertlion commented on features/journalist_can_create_article.feature:
```diff
@@ -0,0 +1,20 @@
+    I would like to able to create an article and choose the right category for it.
+
+  Background:
+    Given the following categories exist
```
you should probably load more than 1 category, and test it

angiecstn replied:
Added.

rupertlion commented on features/support/env.rb:
```diff
@@ -11,7 +11,7 @@
 
 Cucumber::Rails::Database.javascript_strategy = :truncation
-
+Chromedriver.set_version '2.42'
```
Why do we need to set the version?

angiecstn replied:
We added it yesterday with Erik when trying to fix the ChromeDriver troubleshooting, I'll remote it now as it is no longer needed.

rupertlion reviewed (commented):
A few things to think about - also check the spacing/indentation (eg 1/2 tabs vs full tabs)

angiecstn committed:
refactoring and adding more categories to feature test

angiecstn committed:
removing private method in application controller as not needed, refactoring in view to get rid of doubles

ESteinman commented on features/journalist_can_create_article.feature:
```diff
@@ -0,0 +1,20 @@
+    And I fill in "Byline" with "Angelica the Great"
+    And I fill in "Body" with "Angelica in craftacademy"
+    And I select 'Politics' 
+    Then I click "Submit"
```
You should probably test for the notice and maybe if it's actually saved... like:

```
And("there should be a Article with header {string} in the database") do |expected_header|
  expect(Article.last.header).to eq expected_header
end 
```

angiecstn replied:
Not sure whether in this situation it is best practice to put it back or leave it as it is.

ESteinman commented on app/views/layouts/application.html.haml:
```diff
@@ -1,12 +1,16 @@
+  = flash[:error]
+  = yield
 
+  development
```
This shall not be here.

angiecstn replied:
Removed it

angiecstn committed:
refactoring and getting rid of unused code

angiecstn committed:
adding second scenario to test for another choosen category

angiecstn committed:
Merge branch 'development' into create-articles

tochman reviewed (changes requested):
Not really ready for merge. Will do a compleate review during office hours.

tochman committed:
refactors create article scenario
formats code using rubo cop

tochman committed:
separates the main site from admin dashboard

tochman committed:
adds admin controllers and create action
cleans up and removes unused code in main controllers and view
adds new index view for admin dashboard

tochman committed:
more code cleaning

rupertlion commented:
Would the admin dashboard approach in the editor PR have worked if had been completed before this? It has a namespace for admin, with an admin dashboard and an index for articles requiring approval. I was expecting that we would then change the journalist creation view into the admin folder on merging.
